fix: 控制判断时 must_not应该在nested外面 --bug=142983258

Zhaoyikaiii opened the pull request.

Zhaoyikaiii committed:
fix: 控制判断时 must_not应该在nested外面

shamcleren committed:
Merge branch 'master' into fix/must_not_outer_nested/#1010158081142983258

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -942,7 +947,13 @@ func (f *FormatFactory) Query(allConditions metadata.AllConditions) (elastic.Que
+					} else {
+						// 非嵌套字段直接使用MustNot
+						q = f.getQuery(MustNot, queries...)
+					}
```
这一段的作用是什么？

Zhaoyikaiii replied:
已修改。这里是在处理Keyword或者text类型的字段。
如果在遇到value为“”，操作符为:ConditionNotEqual，ConditionNotContains，ConditionNotRegEqual
需要处理成： MustNot -> nested -> field

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -886,7 +887,11 @@ func (f *FormatFactory) Query(allConditions metadata.AllConditions) (elastic.Que
 							case structured.ConditionNotEqual, structured.Ncontains:
-								q = f.getQuery(Must, query)
+								if nf != "" {
+									isNestedBefore = true
```
这里可以利用 nf = "" 的操作来代替 isNestedBefore 的标记位？

Zhaoyikaiii replied:
不可以。这里添加isNestedBefore这个标记位的原因是在最后: 
```go
QE:
			// Add to the appropriate query collection
			if q != nil {
				if nf != "" && !isNestedBefore {
					nestedFields.Add(nf)
					nestedQueries[nf] = append(nestedQueries[nf], q)
				} else {
					nonNestedQueries = append(nonNestedQueries, q)
				}
			}
		}
```

会把当前的q都插入进来。我们现在希望的是能让MustNot在Nested之前插入，所以在前面会构建好一个Nested避免在后面被插入。

Zhaoyikaiii replied:
并且当前这个q会被当成nonNestedQueries处理

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -953,7 +964,7 @@ func (f *FormatFactory) Query(allConditions metadata.AllConditions) (elastic.Que
 			// Add to the appropriate query collection
 			if q != nil {
-				if nf != "" {
+				if nf != "" && !isNestedBefore {
```
这种写法，如果遇到相同的 nested 字段，命中 must 和 must_not 的时候是不是就会导致下面做的分组功能失效？

Zhaoyikaiii replied:
是这个方式吗：
```go
	"nested_must_not_query_mix": {
			conditions: metadata.AllConditions{
				[]metadata.ConditionField{
					{
						DimensionName: "nested1.key",
						Operator:      structured.ConditionNotEqual,
						Value:         []string{"11"},
					},
					{
						DimensionName: "nested1.key",
						Operator:      structured.ConditionNotEqual,
						Value:         []string{""},
					},
				},
			},
			expected: `{"query":{"bool":{"must":[{"bool":{"must_not":{"nested":{"path":"nested1","query":{"match_phrase":{"nested1.key":{"query":"11"}}}}}}},{"bool":{"must_not":{"nested":{"path":"nested1","query":{"match_phrase":{"nested1.key":{"query":""}}}}}}}]}}}`,
		},
```

Zhaoyikaiii committed:
refactor: es处理嵌套字段value为空查询操作符为不等于NEQ

- 如果是Keyword或者text构造的顺序为:MustNot->Nested->Field
- 其他的(数字，布尔类型)构造顺序为:Nested -> Exist -> Field

Zhaoyikaiii committed:
test: 补充测试用例

Zhaoyikaiii committed:
test: 补充测试用例

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -886,7 +887,15 @@ func (f *FormatFactory) Query(allConditions metadata.AllConditions) (elastic.Que
 							case structured.ConditionNotEqual, structured.Ncontains:
-								q = f.getQuery(Must, query)
+								// 813行已经把isExistsQuery 设为 false。所以在这里处理的是非KeyWord和Text类型的字段
+								// 如果是非KeyWord和Text类型字段: nested -> exist -> field
```
上面还有一个 case structured.ConditionNotExisted:，不处理吗？如果不处理的话，就变成 nested -> must_not -> existed -> field 了

Zhaoyikaiii replied:
已修复

Zhaoyikaiii committed:
Merge branch 'master' into fix/must_not_outer_nested/#1010158081142983258

Zhaoyikaiii committed:
feat: 增加对嵌套字段的非存在查询支持

- 在查询构造中，新增对嵌套字段的处理逻辑，支持 MustNot 条件下的嵌套查询。
- 补充相应的单元测试用例以验证新功能。

shamcleren commented on pkg/unify-query/tsdb/elasticsearch/format.go:
```diff
@@ -860,11 +860,21 @@ func (f *FormatFactory) Query(allConditions metadata.AllConditions) (elastic.Que
 				q = elastic.NewExistsQuery(key)
 			case structured.ConditionNotExisted:
-				q = f.getQuery(MustNot, elastic.NewExistsQuery(key))
+				if nf != "" {
```
nf 改个名字，增加可读性，使用 nestedPath

Zhaoyikaiii replied:
已修改

Zhaoyikaiii committed:
fix: 统一嵌套字段变量命名

- 将变量名从 `nf` 修改为 `nestedPath`，提高代码可读性。
- 更新相关逻辑以使用新的变量名，确保查询构造的正确性。

Zhaoyikaiii committed:
Merge branch 'master' into fix/must_not_outer_nested/#1010158081142983258
